### PR TITLE
Update testing documentation

### PR DESCRIPTION
**Pull request by amityadav0**

### CODE REVIEW

File: docs/testing-teal.md [hunk 1]
```diff
@@ -13,29 +13,58 @@
 - [Parser](../packages/runtime/src/parser): Reads TEAL code and converts it to a list of opcodes which are executable by the interpreter. If any opcode/data in teal code is invalid, parser will throw an error.
 - [Interpreter](../packages/runtime/src/interpreter): Executes the list of opcodes returned by parser and updates stack after each execution. At the end of execution, if the stack contains a single non-zero uint64 element then the teal code is approved, and transaction can be executed.
 
+## Block Rounds/Height
+In runtime User can decide the round for each transaction or group of transactions. By default round is set to `2`.
+Block rounds for runtime can be set by using `runtime.setRound(round)`. We can retrieve the round by using `runtime.getRound()` function.<br />
```
Comment (ratik21) [inline docs/testing-teal.md:18]:
```suggestion
Block rounds for runtime can be set by using `runtime.setRound(round)`. We can retrieve the current round by using `runtime.getRound()` function.<br />
```
```suggestion
Block rounds for runtime can be set by using `runtime.setRound(round)`. We can retrieve the round by using `runtime.getRound()` function.<br />
```

Comment (ratik21) [reply docs/testing-teal.md:18]: Why `<br />` ? Should we use them in markdown?

File: docs/testing-teal.md [hunk 2]
```diff
@@ -13,29 +13,58 @@
...
+Example:
+
+      runtime.setRound(5);
```
Comment (ratik21) [inline docs/testing-teal.md:21]:
```suggestion
      runtime.setRound(5); // set current round to 5
```

File: docs/testing-teal.md [hunk 3]
```diff
@@ -13,29 +13,58 @@
...
+This means that current round is set to 5 and transaction will pass only if first valid round is less than 5 and last valid round is greater than 5 for that particular transaction.
+Note: Block round remains same until user changes it by using `runtime.setRound(round)`.
+
+## Flow of Testing
+In this section we will describe the flow of testing:
```
Comment (ratik21) [inline docs/testing-teal.md:26]:
```suggestion
In this section we will describe the flow of testing smart contracts in runtime:
```

File: docs/testing-teal.md [hunk 4]
```diff
@@ -13,29 +13,58 @@
...
+- Prepare Accounts: First of all we will create accounts using `StoreAccount`.
+
+      const john = new StoreAccount(initialAlgo);
```
Comment (ratik21) [inline docs/testing-teal.md:29]: Can we explain the `initialAlgo`/`minBalance` concept here?

File: docs/testing-teal.md [hunk 5]
```diff
@@ -13,29 +13,58 @@
...
+      const bob = new StoreAccount(initialAlgo);
+- Prepare Runtime: After creating accounts we will create a runtime object with those accounts.
+
+      const runtime = new Runtime([john, bob]);
+- Set Rounds: Now we will set the rounds according to our requirement.
+
+      runtime.setRound(20);
+    This means that current round is set to 20 and transactions will pass only if their round range includes 20.
```
Comment (ratik21) [inline docs/testing-teal.md:37]: We already explained this in the above section. I think adding a comment // should be enough.

File: docs/testing-teal.md [hunk 6]
```diff
@@ -13,29 +13,58 @@
...
+- Create Apps/Assets: At this point we have set up everything, now we can create apps or assets or both and test our smart contracts. If we want to create an app, we can use `runtime.addApp()` funtion, similarly for asset we can use `runtime.addAsset()` function.
```
Comment (ratik21) [inline docs/testing-teal.md:38]:
```suggestion
- Create Apps/Assets: At this point we have set up everything, now we can create apps and/or assets and begin testing our smart contracts (present in `/fixtures/project/assets`). For creating a stateful application, we can use `runtime.addApp()` funtion, similarly for creating a new asset we can use `runtime.addAsset()` function.
```

File: docs/testing-teal.md [hunk 7]
```diff
@@ -13,29 +13,58 @@
...
+- Create and Execute Transactions: Now we will create transactions to test functioning of our created apps or smart contracts. Example: Payment Transaction, Atomic Transfers, Asset Transfer etc. we can use `runtime.executeTx()` funtion to execute transaction.
+- State: Please note after a transaction is executed the state of account will be updated, so we will have to update our accounts with the latest state from runtime. We update our accounts using `syncAccounts()` funtion.
```
Comment (ratik21) [inline docs/testing-teal.md:40]:
```suggestion
- Update/Refresh State: Please note after a transaction is executed the state of account will be updated, so we will have to update our accounts with the latest state from runtime. We update our accounts using `syncAccounts()` funtion.
```

File: docs/testing-teal.md [hunk 8]
```diff
@@ -13,29 +13,58 @@
...
+- Verify State: Now, we can verify our state, we will assert if updated state is correct.
```
Comment (ratik21) [inline docs/testing-teal.md:41]: Please add user can use `runtime.getGlobal`, `runtime.getLocal`,.. state functions to verify/assert state

File: docs/testing-teal.md [hunk 9]
```diff
@@ -13,29 +13,58 @@
...
+
 ## Run tests
 In this section we will demonstrate executing transactions with stateless and stateful teal.
 
 ### Stateless TEAL
 
 Let's try to execute a transaction where a user (say `john`) can withdraw funds from an `escrow` account based on a stateless smart contract logic. TEAL code can be found [here](../packages/runtime/test/fixtures/escrow-account/assets/escrow.teal).
-- First let's set up the state: initialize accounts and set up runtime (snippet from mocha test is also provided below).
+- First let's set up the state: initialize accounts, get the logic signature for escrow using `runtime.getLogicSig()` function and set up runtime (snippet from mocha test is also provided below).
```
Comment (ratik21) [inline docs/testing-teal.md:49]: Should we divide stateless TEAL section into contract account (escrow here) and delegated signature?

File: docs/testing-teal.md [hunk 10]
```diff
@@ -117,6 +146,12 @@ Now, we will execute a transaction with stateful TEAL (which increments a global
 
 Complete test can be found [here](../packages/runtime/test/integration/stateful-counter.ts).
 
+## Best Practices
+- Follow Flow of Testing section to set up your tests.
+- Keep each test independent of each other.
+- Use `beforeEach` and `AfterEach` functions to set clear boundaries while testing.
```
Comment (ratik21) [inline docs/testing-teal.md:152]: Can we add `beforeAll`, `afterAll` as well?

Comment (robert-zaremba) [inline docs/testing-teal.md:152]:
```suggestion
- Use `beforeEach` and `afterEach` functions to set clear boundaries while testing.
```

File: docs/testing-teal.md (hunk 6 above)
Comment (amityadav0) [reply docs/testing-teal.md:38]: what do you mean my runtime is read ? @ratik21

File: docs/testing-teal.md (hunk 10 above)
Comment (amityadav0) [reply docs/testing-teal.md:153]: yes

File: docs/testing-teal.md (hunk 4 above)
Comment (amityadav0) [reply docs/testing-teal.md:29]: yes

File: docs/testing-teal.md (hunk 5 above)
Comment (amityadav0) [reply docs/testing-teal.md:37]: removing

File: docs/testing-teal.md (hunk 8 above)
Comment (amityadav0) [reply docs/testing-teal.md:41]: yes

File: docs/testing-teal.md (hunk 6 above)
Comment (amityadav0) [reply docs/testing-teal.md]: ah, you mean `ready`

File: docs/testing-teal.md (hunk 9 above)
Comment (amityadav0) [reply docs/testing-teal.md:50]: you mean add another example, yeah that will be nice to have.

File: docs/testing-teal.md [hunk 11]
```diff
@@ -13,61 +13,159 @@
...
+This means that current round is set to 5 and transaction will pass only if its' first valid round is less or equal 5 and the last valid round is greater than 5.
+Note: Block round remains same until user changes it by calling `runtime.setRound(round)`.
+
+## Flow of Testing
+In this section we will describe the flow of testing smart contracts in runtime:
+- Prepare Accounts: First of all we will create accounts using `StoreAccount`.
+
+      const john = new StoreAccount(initialAlgo);
+      const bob = new StoreAccount(initialAlgo);
+  `initialAlgo`: To set up accounts we can pass the amount of algos we want to have in it.
```
Comment (ratik21) [inline docs/testing-teal.md:31]:
```suggestion
  `initialAlgo`: To set up accounts we can pass the initial amount of algos we want to have in it. It's recommended to have some initial algos (to cover transaction fees and to maintain minimum balance for an account)
```

File: docs/testing-teal.md [hunk 12]
```diff
@@ -13,61 +13,159 @@
...
+
+      // execute transaction
+      await runtime.executeTx(txnParams);
+
+      // check final state (updated accounts)
+      assert.equal(getAcc(runtime, escrow).balance(), initialEscrowHolding - 100); // check if 100 microAlgo's are withdrawn
+      assert.equal(getAcc(runtime, john).balance(), initialJohnHolding + 100);
+    });
+    ```
+    In test, we are first checking the initial balance which we set during initialization. Then we are executing transaction by passing `txnParams`. After execution, we are verifying the account balances if the funds are withdrawn from `escrow`.
```
Comment (ratik21) [inline docs/testing-teal.md:85]:
```suggestion
    In this test, we are first checking the initial balance which we set during initialization. Then we are executing transaction by passing `txnParams`. After execution, we are verifying the account balances if the funds are withdrawn from `escrow`.
```

File: docs/testing-teal.md [hunk 13]
```diff
@@ -131,15 +236,19 @@ Currently, `runtime` supports:
   + `delete` application
   + `closeout` from an application
   + `clearState` of application
+  + `create` an asset
+  + `opt-in` to asset
+  + `transfer` an asset
 
 - Full transaction processing for type `payment`, `application call`
```
Comment (ratik21) [inline docs/testing-teal.md:243]: I think now we support full transaction processing for all types. Can we remove/modify this?

Comment (amityadav0) [reply docs/testing-teal.md:243]: no, we don't for all. we will update it after asset support (revoke, freeze, etc) gets merged.

Comment (ratik21) [reply docs/testing-teal.md:243]: ah, yes. It's not merged. We can update there.

File: docs/testing-teal.md [hunk 14]
```diff
@@ -117,6 +215,13 @@ Now, we will execute a transaction with stateful TEAL (which increments a global
 
 Complete test can be found [here](../packages/runtime/test/integration/stateful-counter.ts).
 
+## Best Practices
+- Follow Flow of Testing section to set up your tests.
+- Structure tests by the AAA pattern: Structure your tests with 3 well-separated sections: Arrange, Act & Assert (AAA). The first part includes the test setup, then the execution of the unit under test, and finally the assertion phase. Following this structure guarantees that the reader spends no brain CPU on understanding the test plan.
```
Comment (robert-zaremba) [inline docs/testing-teal.md:220]:
```suggestion
- Structure tests using AAA pattern: Arrange, Act & Assert (AAA). The first part includes the test setup, then the execution of the unit under test, and finally the assertion phase. Following this structure guarantees that the reader will quickly understand the test plan.
```

**Review by robert-zaremba — APPROVED** [review]
Pre approving. Added comments and suggestions.

**Pull request merged**